sdk/monitor/OpenTelemetry.Exporter.AzureMonitor/src/AzureMonitorTransmitter.cs

```diff
@@ -82,22 +84,59 @@ internal async ValueTask<int> AddBatchActivityAsync(Batch<Activity> batchActivit
         private static TelemetryItem GeneratePartAEnvelope(Activity activity)
         {
             TelemetryItem telemetryItem = new TelemetryItem(PartA_Name_Mapping[activity.GetTelemetryType()], activity.StartTimeUtc);
-            // TODO: Validate if Azure SDK has common function to generate role instance
-            telemetryItem.Tags[ContextTagKeys.AiCloudRoleInstance.ToString()] = "testRoleInstance";
-
+            GetRoleInfo(activity.GetResource(), out var roleName, out var roleInstance);
+            telemetryItem.Tags[ContextTagKeys.AiCloudRole.ToString()] = roleName;
+            telemetryItem.Tags[ContextTagKeys.AiCloudRoleInstance.ToString()] = roleInstance;
             telemetryItem.Tags[ContextTagKeys.AiOperationId.ToString()] = activity.TraceId.ToHexString();
             if (activity.Parent != null)
             {
                 telemetryItem.Tags[ContextTagKeys.AiOperationParentId.ToString()] = activity.Parent.SpanId.ToHexString();
             }
-
-            // TODO: "ai.location.ip"
             // TODO: Handle exception
             telemetryItem.Tags[ContextTagKeys.AiInternalSdkVersion.ToString()] = SdkVersionUtils.SdkVersion;
 
             return telemetryItem;
         }
 
+        internal static void GetRoleInfo(Resource resource, out string roleName, out string roleInstance)
+        {
+            if (resource == null)
+            {
+                roleName = null;
+                roleInstance = null;
+                return;
+            }
+
+            string serviceName = null;
+            string serviceNamespace = null;
+            roleInstance = null;
+
+            foreach (var attribute in resource.Attributes)
+            {
+                if (attribute.Key == Resource.ServiceNameKey && attribute.Value is string)
+                {
+                    serviceName = attribute.Value.ToString();
+                }
+                else if (attribute.Key == Resource.ServiceNamespaceKey && attribute.Value is string)
+                {
+                    serviceNamespace = attribute.Value.ToString();
+                }
+                else if (attribute.Key == Resource.ServiceInstanceIdKey && attribute.Value is string)
+                {
+                    roleInstance = attribute.Value.ToString();
+                }
```

Comment:
i see that you're doing the `attribute.Value is string` three times.
consider putting that in an outer if statement and the key compare in an inner if.